### PR TITLE
Remove some hardcoded trait assumptions from HackyAI.

### LINKED ISSUE #10496: AI won't deploy Conyard if the actor is a helicopter.
A reminder asked by pchote, please let me know if an example code is needed or anyhing.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29810549-ai-won-t-deploy-conyard-if-the-actor-is-a-helicopter?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Fixes #10496 and adds a little bit more sanity to HackyAI.

### CODE REVIEW

File: OpenRA.Mods.Common/AI/HackyAI.cs [hunk 1]
```diff
@@ -133,11 +150,15 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
...
 		[Desc("Tells the AI what unit types fall under the same common name. Only supported entry is Mcv.")]
-		public readonly Dictionary<string, HashSet<string>> UnitsCommonNames = null;
+		[FieldLoader.LoadUsing("LoadUnitCategories")]
+		[FieldLoader.Require]
+		public readonly UnitCategories UnitsCommonNames;
 
 		[Desc("Tells the AI what building types fall under the same common name.",
 			"Possible keys are ConstructionYard, Power, Refinery, Silo , Barracks, Production, VehiclesFactory, NavalProduction.")]
-		public readonly Dictionary<string, HashSet<string>> BuildingCommonNames = null;
+		[FieldLoader.LoadUsing("LoadBuildingCategories")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: Any reason not to put these on the same line?

Comment (pchote) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: Not really.  Is there any reason to put them on the same line?  I don't think we have a convention/style decided for this.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: We don't have a convention but we could start one here.
Also no real reason other than taking up less space.

**Comment by abcdefg30** [comment]
Crashes for me:
```
Red Alert Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.NotImplementedException`: FieldLoader: Missing field `[Type] ` on `UnitCategories`
   at OpenRA.FieldLoader.<.cctor>b__b(String s, Type f) in OpenRA\OpenRA.Game\FieldLoader.cs:Line 63.
   at OpenRA.FieldLoader.GetValue(String fieldName, Type fieldType, MiniYaml yaml, MemberInfo field) in OpenRA\OpenRA.Game\FieldLoader.cs:Line 543.
   at OpenRA.FieldLoader.TryGetValueFromYaml(String yamlName, FieldInfo field, Dictionary`2 md, Object& ret) in OpenRA\OpenRA.Game\FieldLoader.cs:Line 122.
   at OpenRA.FieldLoader.Load(Object self, MiniYaml my) in OpenRA\OpenRA.Game\FieldLoader.cs:Line 99.
   at OpenRA.ActorInfo.LoadTraitInfo(ObjectCreator creator, String traitName, MiniYaml my) in OpenRA\OpenRA.Game\GameRules\ActorInfo.cs:Line 110.
   at OpenRA.ActorInfo..ctor(ObjectCreator creator, String name, MiniYaml node, Dictionary`2 allUnits) in OpenRA\OpenRA.Game\GameRules\ActorInfo.cs:Line 51.
   at OpenRA.RulesetCache.<Load>b__0(MiniYamlNode k, Dictionary`2 y) in OpenRA\OpenRA.Game\GameRules\RulesetCache.cs:Line 64.
   at OpenRA.RulesetCache.<>c__DisplayClass15`1.<LoadYamlRules>b__10(MiniYamlNode wkv, Dictionary`2 wyy) in OpenRA\OpenRA.Game\GameRules\RulesetCache.cs:Line 113.
   at OpenRA.RulesetCache.<>c__DisplayClass15`1.<LoadYamlRules>b__14(MiniYamlNode kv) in OpenRA\OpenRA.Game\GameRules\RulesetCache.cs:Line 121.
   at OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, String debugName, Func`2 logKey, Func`2 logValue) in OpenRA\OpenRA.Game\Exts.cs:Line 366.
   at OpenRA.RulesetCache.LoadYamlRules[T](Dictionary`2 itemCache, String[] files, List`1 nodes, Func`3 f) in OpenRA\OpenRA.Game\GameRules\RulesetCache.cs:Line 121.
   at OpenRA.RulesetCache.Load(Map map) in OpenRA\OpenRA.Game\GameRules\RulesetCache.cs:Line 62.
   at OpenRA.ModData.<.ctor>b__3() in OpenRA\OpenRA.Game\ModData.cs:Line 76.
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at OpenRA.ModData.get_DefaultRules() in OpenRA\OpenRA.Game\ModData.cs:Line 38.
   at OpenRA.Map.<PostInit>b__f() in OpenRA\OpenRA.Game\Map\Map.cs:Line 460.
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at OpenRA.Map.PreloadRules() in OpenRA\OpenRA.Game\Map\Map.cs:Line 583.
   at OpenRA.ModData.PrepareMap(String uid) in OpenRA\OpenRA.Game\ModData.cs:Line 183.
   at OpenRA.Game.StartGame(String mapUID, WorldType type) in OpenRA\OpenRA.Game\Game.cs:Line 154.
   at OpenRA.Game.LoadShellMap() in OpenRA\OpenRA.Game\Game.cs:Line 443.
   at OpenRA.Mods.Common.LoadScreens.BlankLoadScreen.StartGame(Arguments args)
   at OpenRA.Game.InitializeMod(String mod, Arguments args) in OpenRA\OpenRA.Game\Game.cs:Line 399.
   at OpenRA.Game.Initialize(Arguments args) in OpenRA\OpenRA.Game\Game.cs:Line 288.
   at OpenRA.Program.Run(String[] args) in OpenRA\OpenRA.Game\Support\Program.cs:Line 115.
   at OpenRA.Program.Main(String[] args) in OpenRA\OpenRA.Game\Support\Program.cs:Line 39.

**Comment by pchote** [comment]
@abcdefg30: can you please give steps to reproduce? The RA mod and its AI works fine for me on this branch.

**Comment by pchote** [comment]
Fixed.  Note for future reference: `FieldLoader.Require` and `FieldLoader.LoadUsing` both subclass the same base attribute type, which makes them mutually incompatible(!).  This took far too long to track down.

**Comment by abcdefg30** [comment]
Looks good to me now. :+1: / :white_check_mark:

**Comment by Mailaender** [comment]
Bots will still build a new MCV and deploy it (after quite an odyssey in Dune 2000) for replacement. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a5dbd0b73aca11281acddde727ff0d41da39bfdf...c57bdb69e4791b95ff3123c8473a68103a55bbf9)

**Pull request merged**